src/game/plugin/pluginmanager.cpp

```diff
@@ -21,45 +21,269 @@
+        PluginInfo info;
+        Library lib;
+        Status stat;
+    };
+
+public:
+    PluginLoader() {
+		constexpr const char* pathSep =
+#if (BOOST_OS_WINDOWS)
+			";";
+#else
+			":";
+#endif
+		std::string env = std::getenv("PATH");
+		env += pathSep + filesystem::absolute("./Modules/").string() + pathSep;
+		env = "PATH=" + env;
+		char* nenv = new char[env.size() + 1];
+		std::strcpy(nenv, env.c_str());
+		putenv(nenv);
+        walk();
+        for (auto&& x : mMap)
+            loadPlugin(x.second);
+    }
+
+    void loadPlugin(const std::string uri) { loadPlugin(mMap[uri]); }
+
+    void loadPlugin(LoadingInfo& inf) noexcept {
+        if (inf.stat != Status::Pending) return;
+        try {
+            infostream << "Loading Module: " << inf.info.uri;
+            for (auto& x : inf.info.dependencies) {
+                loadPlugin(x.uri);
+                try {
+                    verify(x);
+                }
+                catch (std::exception& e) {
+                    warningstream << "Module Denpendency " << x.uri << " Of: " << inf.info.uri <<
+                        " Failed For: " << e.what();
+                    if (x.isOptional)
+                        warningstream << "Dependency Skipped For It Is Optional";
+                    else
+                        throw;
+                }
+            }
+			std::unique_ptr<PluginObject> object;
+			switch (inf.info.type) {
+		    case ModuleType::C:
+				object = std::make_unique<DefaultPluginObject>(inf.lib);
+				break;
+			case ModuleType::CPP:
+			{
+				const auto getObject = inf.lib.get<PluginObject* NWAPICALL()>("nwModuleGetObject");
+				if (getObject)
+					object.reset(getObject());
+				else
+					throw std::runtime_error("Module Has no nwModuleGetObject function, skipping finalization!");
+			}
+				break;
+			default:
+				break;
+			}
+            // No Error, Load Success
+            inf.stat = Status::Success;
+			mResult.map[inf.info.uri] = mResult.modules.size();
+			PluginPair pair{ std::move(inf.lib), std::move(object) };
+			mResult.modules.push_back(std::move(pair));
+        }
+        catch (std::exception& e) {
+            warningstream << "Module: " << inf.info.uri << " Failed For: " << e.what();
+            inf.stat = Status::Fail;
+        }
+        catch (...) {
+            warningstream << "Module: " << inf.info.uri << " Failed For Unknown Reason";
+            inf.stat = Status::Fail;
+        }
+    }
+
+    void verify(DependencyInfo& inf) {
+        auto& depStat = mMap[inf.uri];
+        if (depStat.stat != Status::Success) throw std::runtime_error("Dependency Load Failure");
+        if (inf.vRequired > depStat.info.thisVersion) throw std::runtime_error("Dependency Version Mismatch : Version Unsupported");
+        if (inf.vRequired < depStat.info.conflictVersion)
+            warningstream << "Dependency Version Mismatch : Confilict. Module May Behave Strangely. Use with care";
+    }
+
+    void walk() {
```

Comment:
Any reason to put `walk()` as a public function? The name of  `walk` is too vague to be a public function. Consider either making it private or renaming it.